File: elixir/aoc/lib/aoc/day06a.ex
defmodule AOC.Day06a do
  use AOC

  def solution(path) do
    [times, distances] = path |> read_lines() |> Enum.map(&String.trim/1)
    [_, times] = times |> String.split(":", trim: true)
    times = times |> String.split(" ", trim: true) |> Enum.map(&String.to_integer/1)
    [_, distances] = distances |> String.split(":", trim: true)
    distances = distances |> String.split(" ", trim: true) |> Enum.map(&String.to_integer/1)
    races = Enum.zip(times, distances)

    races
    |> Enum.map(&winning_strategies/1)
    |> Enum.reduce(1, &(&1 * &2))
  end

  def winning_strategies({time, distance}) do
    1..time
    |> Enum.filter(&winning?(&1, time, distance))
    |> Enum.count()
  end

  def winning?(holding, time, distance) do
    holding * (time - holding) > distance
  end
end

File: ruby/aoc/lib/aoc/day06a.rb
module AOC
  class Day06a < Day
    def winning_combinations
      parse_input.map(&method(:winning_strategies)).reduce(1, &:*)
    end

    def winning_strategies(constraints)
      time, distance = constraints
      (1..time).select do |hold|
        hold * (time - hold) > distance
      end.count
    end

    def parse_input
      times, distances = File.readlines(@path, chomp: true)
      _, times = times.split(': ')
      times = times.split(" ").map(&:to_i)
      _, distances = distances.split(': ')
      distances = distances.split(" ").map(&:to_i)
      times.zip(distances)
    end
  end
end
